File: src/auth/reauthSession.ts
/**
 * Re-auth Session Store
 *
 * Short-lived session for re-auth challenge/response correlation.
 * TTL = REAUTH_SESSION_TTL_SECONDS (90s).
 *
 * Redis key pattern: reauth_session:{sessionId}
 * Distinct from eap_session: used for full auth.
 */

import crypto from 'node:crypto';
import { redis } from '../db/redis.js';
import { EAP_AKA } from '../config/constants.js';

export interface ReauthSessionData {
  reauthId: string;
  nextReauthId: string;
  nonceS: string;      // base64
  counter: number;
  identifier: number;
  kAut: string;         // base64
  kEncr: string;        // base64
  mk: string;           // base64
  subscriberId: string;
  imsi: string;
}

function sessionKey(sessionId: string): string {
  return `reauth_session:${sessionId}`;
}

/**
 * Create a new re-auth session. Returns the generated session ID.
 */
export async function createReauthSession(data: ReauthSessionData): Promise<string> {
  const sessionId = crypto.randomUUID();
  const key = sessionKey(sessionId);

  const fields: Record<string, string> = {
    reauthId: data.reauthId,
    nextReauthId: data.nextReauthId,
    nonceS: data.nonceS,
    counter: String(data.counter),
    identifier: String(data.identifier),
    kAut: data.kAut,
    kEncr: data.kEncr,
    mk: data.mk,
    subscriberId: data.subscriberId,
    imsi: data.imsi,
  };

  await redis.hset(key, fields);
  await redis.expire(key, EAP_AKA.REAUTH_SESSION_TTL_SECONDS);

  return sessionId;
}

/**
 * Retrieve a re-auth session by session ID.
 * Returns null if not found or expired.
 */
export async function getReauthSession(sessionId: string): Promise<ReauthSessionData | null> {
  const key = sessionKey(sessionId);
  const data = await redis.hgetall(key);

  if (!data || Object.keys(data).length === 0) {
    return null;
  }

  return {
    reauthId: data.reauthId!,
    nextReauthId: data.nextReauthId!,
    nonceS: data.nonceS!,
    counter: parseInt(data.counter!, 10),
    identifier: parseInt(data.identifier!, 10),
    kAut: data.kAut!,
    kEncr: data.kEncr!,
    mk: data.mk!,
    subscriberId: data.subscriberId!,
    imsi: data.imsi!,
  };
}

/**
 * Delete a re-auth session.
 */
export async function deleteReauthSession(sessionId: string): Promise<void> {
  await redis.del(sessionKey(sessionId));
}

File: internal/eapaka/reauth_session.go
package eapaka

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/google/uuid"
	"github.com/hkassaei/entitlements-as-a-service-poc-go/internal/config"
	"github.com/redis/go-redis/v9"
)

// ReauthSessionData represents a re-auth session for challenge/response correlation.
type ReauthSessionData struct {
	ReauthID     string
	NextReauthID string
	NonceS       string // base64
	Counter      int
	Identifier   int
	KAut         string // base64
	KEncr        string // base64
	MK           string // base64
	SubscriberID string
	IMSI         string
}

// ReauthSessionStore manages re-auth sessions in Redis.
type ReauthSessionStore struct {
	client *redis.Client
}

// NewReauthSessionStore creates a new ReauthSessionStore.
func NewReauthSessionStore(client *redis.Client) *ReauthSessionStore {
	return &ReauthSessionStore{client: client}
}

func reauthSessionKey(sessionID string) string {
	return "reauth_session:" + sessionID
}

// Create creates a new re-auth session and returns the session ID.
func (s *ReauthSessionStore) Create(ctx context.Context, data *ReauthSessionData) (string, error) {
	sessionID := uuid.New().String()
	key := reauthSessionKey(sessionID)

	fields := map[string]interface{}{
		"reauthId":     data.ReauthID,
		"nextReauthId": data.NextReauthID,
		"nonceS":       data.NonceS,
		"counter":      strconv.Itoa(data.Counter),
		"identifier":   strconv.Itoa(data.Identifier),
		"kAut":         data.KAut,
		"kEncr":        data.KEncr,
		"mk":           data.MK,
		"subscriberId": data.SubscriberID,
		"imsi":         data.IMSI,
	}

	pipe := s.client.Pipeline()
	pipe.HSet(ctx, key, fields)
	pipe.Expire(ctx, key, time.Duration(config.ReauthSessionTTLSeconds)*time.Second)
	if _, err := pipe.Exec(ctx); err != nil {
		return "", fmt.Errorf("create reauth session: %w", err)
	}

	return sessionID, nil
}

// Get retrieves a re-auth session by ID. Returns nil if not found.
func (s *ReauthSessionStore) Get(ctx context.Context, sessionID string) (*ReauthSessionData, error) {
	key := reauthSessionKey(sessionID)
	data, err := s.client.HGetAll(ctx, key).Result()
	if err != nil {
		return nil, fmt.Errorf("get reauth session: %w", err)
	}
	if len(data) == 0 {
		return nil, nil
	}

	counter, _ := strconv.Atoi(data["counter"])
	identifier, _ := strconv.Atoi(data["identifier"])

	return &ReauthSessionData{
		ReauthID:     data["reauthId"],
		NextReauthID: data["nextReauthId"],
		NonceS:       data["nonceS"],
		Counter:      counter,
		Identifier:   identifier,
		KAut:         data["kAut"],
		KEncr:        data["kEncr"],
		MK:           data["mk"],
		SubscriberID: data["subscriberId"],
		IMSI:         data["imsi"],
	}, nil
}

// Delete removes a re-auth session.
func (s *ReauthSessionStore) Delete(ctx context.Context, sessionID string) error {
	return s.client.Del(ctx, reauthSessionKey(sessionID)).Err()
}
